signer/kms/azure/signer.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2025 The Witness Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azure
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"fmt"
+	"io"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/security/keyvault/azkeys"
+	"github.com/in-toto/go-witness/cryptoutil"
+	kms "github.com/in-toto/go-witness/signer/kms"
+)
+
+var azureSupportedAlgorithms = []string{
+	string(azkeys.SignatureAlgorithmRS256),
+	string(azkeys.SignatureAlgorithmRS384),
+	string(azkeys.SignatureAlgorithmRS512),
+	string(azkeys.SignatureAlgorithmPS256),
+	string(azkeys.SignatureAlgorithmPS384),
+	string(azkeys.SignatureAlgorithmPS512),
+	string(azkeys.SignatureAlgorithmES256),
+	string(azkeys.SignatureAlgorithmES256K),
+	string(azkeys.SignatureAlgorithmES384),
+	string(azkeys.SignatureAlgorithmES512),
+}
+
+var azureSupportedHashFuncs = []crypto.Hash{
+	crypto.SHA256,
+	crypto.SHA384,
+	crypto.SHA512,
+}
+
+// SignerVerifier is a cryptoutil.SignerVerifier that uses the Azure Key Vault
+type SignerVerifier struct {
+	reference string
+	client    client
+	hashFunc  crypto.Hash
+}
+
+// LoadSignerVerifier generates signatures using the specified key object in Azure Key Vault and hash algorithm.
+func LoadSignerVerifier(ctx context.Context, ksp *kms.KMSSignerProvider) (*SignerVerifier, error) {
+	a := &SignerVerifier{
+		reference: ksp.Reference,
+	}
+
+	var err error
+	a.client, err = newAzureClient(ctx, ksp)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, hashFunc := range azureSupportedHashFuncs {
+		if hashFunc == ksp.HashFunc {
+			a.hashFunc = ksp.HashFunc
+		}
+	}
+
+	if a.hashFunc == 0 {
+		// Default to SHA256 if not specified
+		a.hashFunc = crypto.SHA256
+	}
+
+	return a, nil
+}
+
+// KeyID returns the key identifier for the key used by this signer.
+func (a *SignerVerifier) KeyID() (string, error) {
+	return a.reference, nil
+}
+
+// Sign signs the provided message using Azure Key Vault. If the message is provided,
+// this method will compute the digest according to the hash function specified
+// when the Signer was created.
+func (a *SignerVerifier) Sign(message io.Reader) ([]byte, error) {
+	var err error
+	ctx := context.TODO()
```

Comment:
We've used `context.Background()` in the GCP KMS implementation. https://github.com/in-toto/go-witness/blob/main/signer/kms/gcp/signer.go#L79

Generally speaking, we've not been very good or consistent in how to allow passing context through to signer/verifiers. @mikhailswift did you have any thoughts on this?

Reply:
We definitely need to do a full pass on adding ctxs as parameters of the functions instead of grabbing todo/background. It’ll probably be a bit painful, affect multiple interfaces, and be a large changeset, though. But I think it’s likely the right way to get contexts into the functions that need them

Reply:
Maybe an inexperienced question but does this mean introducing a Context at the 'root' level and passing it along to all the places that could use it? And could this be 'refactored' piece by piece, so the change sets stay manageable?

Reply:
The biggest pain in adding context around would be adding it to interfaces that don't currently have it -- since we'd have to ensure all the current implementations of that interface get updated appropriately. Not a difficult task, just one that could somewhat spiral.

Otherwise yeah, you've got the right of it. For the most part in go-witness we don't need to create a context to pass in, users of the library should be providing it, and we just have to fish it down where it needs to be.